src/main/java/frc/robot/subsystems/conveyor/Conveyor.java

```diff
@@ -0,0 +1,108 @@
+package frc.robot.subsystems.conveyor;
+
+import com.ctre.phoenix.motorcontrol.TalonFXControlMode;
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Ports;
+import frc.robot.utils.DeadbandLaserSensor;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static frc.robot.Constants.Conveyor.*;
+
+public class Conveyor extends SubsystemBase {
+    private static final AtomicInteger balls = new AtomicInteger(Constants.Conveyor.INITIAL_BALLS_AMOUNT);
```

Comment:
Cool, I didn't know about `AtomicInteger`. But shouldn't the `Atmoic*` stuff be used when you have multiple threads that can change the variable?

Reply:
You over-complicate the code and I see no reason for that.
The code is single-threaded, so you are at no risk `balls`'s value will be wrong. Thus, using `synchronized` or `AtomicInteger` is unnecessary.
Moreover, you don't need `static` because this class is a singleton.

In the code for _Asteria_, `balls` wasn't static nor `synchronized` or `AtomicInteger` were used, and the code worked perfectly. In fact, [it was you who've written it](https://github.com/Galaxia5987/robot-2020/blob/0c6325625ff0e0f5544c819be2e98ad6173b071e/src/main/java/frc/robot/subsystems/conveyor/Conveyor.java#L24)…

This is my last comment on the matter unless you wish to discuss it more.

Reply:
You don't see a reason, but I see.
Example, What if we run the `FeedShooter` command, and in the beginning we have 4 balls, and another ball enter into the conveyor (so we need to increment the amount of balls) and by the **same** time, a ball is being taken out, now the amount of balls can be either 3,4 or 5, and the data is lost, and btw, it didn't work as we expected, so we didn't use it in the competition....

Reply:
It will still be OK.
First of all, the chance this scenario will happen is very rare – it means you're intaking and shooting at the same time.
But even if you do so it will be fine. Let's take your example where `balls == 4`, and a Power Cell enters the conveyor while another exists through the shooter. Let's look at the code inside `periodic()`:
https://github.com/Galaxia5987/robot-2021/blob/3332132f23f8062f4c06044016b6644f5b959458/src/main/java/frc/robot/subsystems/conveyor/Conveyor.java#L98-L106
The condition at line 98 will evaluate to `true`, so line 100 will decrement `balls` to `3`.
The condition at line 104 will evaluate to `True`, so line 106 will increment `balls` to `4`.

Again, your code is synchronous. Even if something happens "by the **same** time", it will happen one after the other.

BTW, after looking at the code referenced above, I realized that if a Power Cell _exists_ through the intake while another is being shot, `balls` really will hold a wrong value (which is another rare scenario). See why?
That would happen even if you used `synchronized`, but fortunately, there's an easy fix.

Reply:
Congratulation, You passed the test.
And now the real problem IMO:
Do you think that the `periodic()` function stops when we start running a command? Answer: No, the `periodic()` function runs and doesn't care about the commands. How is that, that the function still running while the commands run? They run the commands and the `periodic()` in other threads...
BTW, I don't need to make everything thread safe, only the `isConveyor...` functions, because only they run on other threads.

P.S, even if the scenario is _rare_ it still can happen, and I should think about it and find a way to handle it.

Reply:
> Congratulation, You passed the test.

![image](https://media.tenor.com/images/d936610c7ab637c0c8c7bed796e6a202/tenor.gif)
> Do you think that the periodic() function stops when we start running a command?

No.
> Answer: No

Reply:
> you know that cores != amount of threads, right?

I'm not talking about the total number of threads, but the total number of **simultaneously running** threads. Unless your CPU uses [Intel Hyper-Threading](https://en.wikipedia.org/wiki/Hyper-threading) or similar technology (in which case you'll count "logical cores"), I don't see a scenario where you can run more threads than the number of cores.
> and I just saw the implementation of periodic and it **still** runs when commands run...

They both might be scheduled to run, but when they actually run is up to the operating system to decide.

> How the hack is it related? if the balls are being called by more than 1 thread then you need to use `synchronized`

No, only if `balls` is **changed** by more than one thread. A race condition, which is the situation you describe here, can happen only if multiple threads try to change shared data at the same time.